### ISSUE TITLE
Simplify engine's script API registration and calling method

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1223]
I wanted to make this ticket *very* long ago, but although seeing a clear problem I could not figure out which solution I prefer most.

The problem that I'd like to address is the script API registration in the engine. This is one of the things that were done very early in open sourced AGS engine as a method to a) make api calls safer, b) react to changes to script interpreter where the values are now stored in self-described RuntimeScriptValue struct instead of being pure ints of unknown origin.
When I was doing this I also took a notice on the work of ScummVM developer who made the first AGS port attempt, and seeing how she invented similar thing encouraged me to go down this path further.

I apologize beforehand for a long read ahead, but I must explain few things and give code examples too. Maybe following is not even enough, in which case I may add further explanations and links to code to make things more clear.

---

### Current implementation

In a nutshell, current system requires you to write a "translator" function that would unwrap values coming from script interpreter and call an actual function.
There are two prototypes - for static and object member script function, - and they look like this:
<pre>
typedef RuntimeScriptValue ScriptAPIFunction(const RuntimeScriptValue *params, int32_t param_count);
typedef RuntimeScriptValue ScriptAPIObjectFunction(void *self, const RuntimeScriptValue *params, int32_t param_count);
</pre>
The example of such "translator" function:
<pre>
RuntimeScriptValue Sc_Object_Animate(void *self, const RuntimeScriptValue *params, int32_t param_count)
{
    API_OBJCALL_VOID_PINT5(ScriptObject, Object_Animate);
}
</pre>

The `API_OBJCALL_VOID_PINT5` is a macro, which unwinds into actual function call. This is done for simplicity sake. There is a huge list of such macros found in [Engine/script/script_api.h](https://github.com/adventuregamestudio/ags/blob/master/Engine/script/script_api.h).
For example, the above `API_OBJCALL_VOID_PINT5` is defined like:
<pre>
#define API_OBJCALL_VOID_PINT5(CLASS, METHOD) \
    ASSERT_OBJ_PARAM_COUNT(METHOD, 5); \
    METHOD((CLASS*)self, params[0].IValue, params[1].IValue, params[2].IValue, params[3].IValue, params[4].IValue); \
    return RuntimeScriptValue((int32_t)0)
</pre>
What happens here is that it calls given method, and passes elements of `params` array. Then returns some value wrapped in a RuntimeScriptValue.

The API functions are registered in the engine using -
<pre>
bool ccAddExternalStaticFunction(const String &name, ScriptAPIFunction *pfn);
bool ccAddExternalObjectFunction(const String &name, ScriptAPIObjectFunction *pfn);
</pre>
In practice registration looks like this:
<pre>
ccAddExternalObjectFunction("Object::Animate^5", Sc_Object_Animate);
ccAddExternalObjectFunction("Object::get_X", Sc_Object_GetX);
ccAddExternalObjectFunction("Object::set_X", Sc_Object_SetX);
</pre>
You may find these registration lists in code, for example this is a list for Object type:
https://github.com/adventuregamestudio/ags/blob/master/Engine/ac/object.cpp#L899

---

### Problem

Current implementation has a benefit of behaving as a argument and return value type check. Because the real function is known, trying to use it with wrong macro (wrong types or number of arguments) would usually cause compilation error.
It however has two major issues.

1. A lot of extra boilerplate code. Basically, for each function you have to add you also have to add extra function that unwraps internal script value types. This is not critical, but obnoxious.
2. The way these "translator" functions are registered, and because they are using internal engine's types not meant for anything else, prevents from sharing same registration with other potential users. To elaborate, consider plugins that also may query engine API, or imagine we decided to support more script languages (either right in the engine or also through plugin api). Even now there has to be separate registration made for script interpreter and plugins (and tbh we were ditching plugin registration for many years, so it's probably missing most new API starting with 3.3.0).

---

### Potential solutions

I must note that the first problem could be reduced by moving "translator" function definition to macros completely.
Let's look at the following example of a "translator" function, we'll see that it already uses macros to simplify its writing:
<pre>
RuntimeScriptValue Sc_Object_GetX(void *self, const RuntimeScriptValue *params, int32_t param_count)
{
    API_OBJCALL_INT(ScriptObject, Object_GetX);
}
</pre>

What could be done instead is this:
<pre>
DEFINE_API_OBJCALL_INT(ScriptObject, Object_GetX)
</pre>
where `DEFINE_API_OBJCALL_INT` would be a macro declared like
<pre>
#define DEFINE_API_OBJCALL_INT(CLASS, METHOD) \
RuntimeScriptValue Sc_##METHOD(void *self, const RuntimeScriptValue *params, int32_t param_count)
{
    API_OBJCALL_INT(CLASS, METHOD);
}
</pre>
Above would allow much smaller function declarations, although perhaps some cases may require manual code. That would look like:
<pre>
DEFINE_API_OBJCALL_VOID_PINT5(ScriptObject, Object_Animate)
DEFINE_API_OBJCALL_INT(ScriptObject, Object_GetX)
DEFINE_API_OBJCALL_PINT(ScriptObject, Object_SetX)
... and so on ...
</pre>

Further optional step to simplify this could be generating these automatically from script api header or some description. Nick Sonneveld proposed this in the past.

Above would solve the problem of developers' convenience.
Unfortunately that still does not resolve the second problem (also, whether created by hand or generated, this extra code will still be part of the engine).

---

What I am looking for instead is the system that would let skip writing extra code per API item, and would allow to reuse same registration for any kind of uses we may have for it.

Over time, the simpliest solution that came to mind, is to instead of registering myriad of "translator" functions, one per real one, have *one translator function* with a switch, and register API functions providing their type description, for instance in a form of enum.
So, for example, instead of having this

<pre>
RuntimeScriptValue Sc_Object_GetX(void *self, const RuntimeScriptValue *params, int32_t param_count)
{
    API_OBJCALL_INT(ScriptObject, Object_GetX);
}

RuntimeScriptValue Sc_Object_SetX(void *self, const RuntimeScriptValue *params, int32_t param_count)
{
    API_OBJCALL_VOID_PINT(ScriptObject, Object_SetX);
}
</pre>

for literally each registered function, we may have something like this:

<pre>
enum APIFuncType
{
    AFT_OBJCALL_INT,
    AFT_OBJCALL_VOID_PINT,
}

RuntimeScriptValue CallAPIFunc(APIFuncType func_type, void *pfunc, void *self,
                                                     const RuntimeScriptValue *params, int32_t param_count)
{
    switch (func_type)
    {
    ...
    case AFT_OBJCALL_INT:          return RuntimeScriptValue().SetInt32(pfunc(self));
    case AFT_OBJCALL_VOID_PINT:    pfunc(self, params[0].IValue); return RuntimeScriptValue((int32_t)0);
    ...
    }
}
</pre>

*NOTE: obviously not calling `pfunc` directly, but casting it to appropriate function prototype.*

### Pros and cons

**Pros**
1. Smaller system, which is perhaps more clear and simplier;
2. **Many times** less boilerplate code, less hassle when adding new API functions;
3. API functions need to be registered only once; provided there's all necessary description same registration info may be used to link engine API to any script interpreter or plugins.

**Cons**
1. Will take an effort to change the code. We should expect new season of engine bugs due to typos and oversights lead to incorrectly linked API.
2. We loose automatic type safety and have to rely on programmers to correctly register function type. This probably is still solveable using certain C++ features.

An example of a very straightforward solution for type safety could be helper function template, where implementations would deduce a type and pass it further to actual registration.
Considering we have something like
<pre>
template<typename FuncProto>
void RegisterAPIFunc(const String &name, FuncProto *pfn) { /* do nothing or error? */ }
</pre>
then we might specialize that as
<pre>
template <>
void RegisterAPIFunc<void (*)(int)>(const String &name, FuncProto *pfn)
{
    RegisterAPIFunc(SCALL_VOID_PINT, name, pfn);
}
</pre>

---

Regardless, this effort, if taken, should be timed for a new development period, not to be along with urgently anticipated changes, as it may take time to make things are stable enough for a release.

**Comment by rofl0r** [comment]
what's the approach taken in the scummvm port ? i didn't look at it for many years...
btw the link to the sourcecode line you have in the description could be made futureproof by replacing master with the commit sha1 hash.

**Comment by ivan-mogilko** [comment]
> what's the approach taken in the scummvm port ? i didn't look at it for many years...

It had similar sort of "translator" function that accepted script values, but some of them had actual implementation in place instead of calling anything further:

https://github.com/adventuregamestudio/scummvm/blob/ags/engines/ags/scripting/object.cpp#L35

By the way, although unrelated, but for the reference, another person completely remade scummvm port using current ags engine code instead, and I heard they already merged it to scummvm master.

**Comment by rofl0r** [comment]
> By the way, although unrelated, but for the reference, another person completely remade scummvm port using current ags engine code instead, and I heard they already merged it to scummvm master.
scummvm/scummvm#2738

wow, thanks for the heads up. that's kinda gross imo, since it appears that the code is mostly copied from ags (ignoring its license) including all its legacy baggage, whereas the fuzzie port was a cleanroom reimplementation, and especially the bytecode vm seemed to have been a lot more efficient and circumvented the pointer/int issues that plagued ags codebase for some years (btw, i myself spent at least 2 week of work on fuzzie's scummvm port too).

**Comment by ivan-mogilko** [comment]
Well, I expressed same opinion to the dev who was working on it, but he said it's simplier to maintain and sync with the latest engine this way.

**Comment by rofl0r** [comment]
ftr the original scummvm port seemed to have used a string containing e.g. "i" for integer arguments and "s" for string arguments to determine types used by script func https://github.com/adventuregamestudio/scummvm/blob/ags/engines/ags/scripting/object.cpp#L975

**Comment by ivan-mogilko** [comment]
Right, and this is how it's called:
https://github.com/adventuregamestudio/scummvm/blob/ags/engines/ags/script.cpp#L1447

**Comment by ericoporto** [comment]
If you have functions named like `Object_Animate` exposed in a header file, and you have an AGS Script header, you could probably parse the AGS Script header and a bunch of cpp headers to generate a big cpp file that has all those calls in it, provided there's some logic in how things are named. 

I actually have been thinking about something like this after doing some big plugins, but in my mind maintaining such tools is more hassle than manually writing these Script glue stuff. 

The template way you mentioned seem like the more maintainable way.

**Comment by ivan-mogilko** [comment]
Expanded the first post with an "Option 2", which is reverse to the solution option mentioned initially (i forgot to do that earlier).


<details> <summary>You may read it in the first post, but I'll add it also here for the quicker reference, under a spoiler.</summary>
Option 2: use new prototype style for both engine and script API

This has been a way of original [ScummVM AGS port attempt](https://github.com/adventuregamestudio/scummvm/tree/ags).

Instead of having a "translator" function redirecting a call to the "actual" engine function, we could have them merged, and have engine API implementation use this (or similar) prototype:

<pre>
typedef RuntimeScriptValue ScriptAPIFunction(const RuntimeScriptValue *params, int32_t param_count);
typedef RuntimeScriptValue ScriptAPIObjectFunction(void *self, const RuntimeScriptValue *params, int32_t param_count);
</pre>

These functions then are registered for both script API and external (plugin etc) API, and may even contain parts of game logic implementation if wanted.
For example, see how it was done here: https://github.com/adventuregamestudio/scummvm/blob/ags/engines/ags/scripting/object.cpp

There's an upside and downside to this.

**Pros**
1. A uniform prototype for all script function implementations (or 2 prototypes: one for a static function and another for a member function).
2. When script doing a function call, there's no need to redirect it once more (like is done in option 1 above).

**Cons**
1. Plugin API that registers custom script functions will have to be adjusted, and plugins that export their own script functions will have to use same prototype.
2. Similarily, any plugins that query engine's script functions for calling them will stop working and will have to be updated.

In a nutshell, this is somewhat easier for engine's VM, but might be annoying for plugin writers.

</details>
